File: python/behavioral/14_command.py
"""
设计模式: 命令模式 (Command)
分类: 行为型

意图: 将请求封装为对象，从而支持参数化、排队、日志、撤销/重做操作。
核心思想: 命令对象封装"动作 + 接收者"，调用者只依赖命令接口。
编程范式: OOP — 封装变化 + 撤销栈；Python 额外展示函数作为命令。
"""

from __future__ import annotations
from abc import ABC, abstractmethod
from collections import deque


# ── 接收者 ────────────────────────────────────────────────────────────────
class TextEditor:
    def __init__(self) -> None:
        self.text = ""

    def append(self, s: str) -> None:   self.text += s
    def delete_last(self, n: int) -> None: self.text = self.text[:-n] if n else self.text
    def print(self) -> None:            print(f'Text: "{self.text}"')


# ── 命令接口 ──────────────────────────────────────────────────────────────
class Command(ABC):
    @abstractmethod
    def execute(self) -> None: ...

    @abstractmethod
    def undo(self) -> None: ...


# ── 具体命令 ──────────────────────────────────────────────────────────────
class AppendCommand(Command):
    def __init__(self, editor: TextEditor, value: str) -> None:
        self._editor = editor
        self._value  = value

    def execute(self) -> None: self._editor.append(self._value)
    def undo(self)    -> None: self._editor.delete_last(len(self._value))


# ── 调用者 ────────────────────────────────────────────────────────────────
class CommandInvoker:
    def __init__(self) -> None:
        self._history: deque[Command] = deque()

    def run(self, cmd: Command) -> None:
        cmd.execute()
        self._history.append(cmd)

    def undo(self) -> None:
        if self._history:
            self._history.pop().undo()


# ── 演示 ──────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    editor  = TextEditor()
    invoker = CommandInvoker()

    invoker.run(AppendCommand(editor, "Hello"))
    invoker.run(AppendCommand(editor, ", World"))
    editor.print()

    invoker.undo(); editor.print()
    invoker.undo(); editor.print()

File: cpp/behavioral/14_command.cpp
/**
 * 设计模式: 命令模式 (Command)
 * 分类: 行为型
 *
 * 意图: 将请求封装为对象，从而支持参数化、排队、日志、撤销/重做操作。
 * 核心思想: 命令对象封装"动作 + 接收者"，调用者只依赖命令接口。
 * 编程范式: OOP — 封装变化 + 命令队列 + 撤销栈。
 */

#include <iostream>
#include <memory>
#include <stack>
#include <string>
#include <vector>

// ── 接收者 ────────────────────────────────────────────────────────────────
struct TextEditor {
    std::string text;
    void append(const std::string& s) { text += s; }
    void deleteLast(size_t n)         { if (n <= text.size()) text.resize(text.size() - n); }
    void print() const                { std::cout << "Text: \"" << text << "\"\n"; }
};

// ── 命令接口 ──────────────────────────────────────────────────────────────
struct Command {
    virtual void execute() = 0;
    virtual void undo()    = 0;
    virtual ~Command() = default;
};

// ── 具体命令 ──────────────────────────────────────────────────────────────
struct AppendCommand : Command {
    TextEditor& editor;
    std::string value;
    AppendCommand(TextEditor& e, std::string v) : editor(e), value(std::move(v)) {}
    void execute() override { editor.append(value); }
    void undo()    override { editor.deleteLast(value.size()); }
};

// ── 调用者（持有命令历史）────────────────────────────────────────────────
class CommandInvoker {
    std::stack<std::unique_ptr<Command>> history_;
public:
    void run(std::unique_ptr<Command> cmd) {
        cmd->execute();
        history_.push(std::move(cmd));
    }
    void undo() {
        if (!history_.empty()) {
            history_.top()->undo();
            history_.pop();
        }
    }
};

int main() {
    TextEditor editor;
    CommandInvoker invoker;

    invoker.run(std::make_unique<AppendCommand>(editor, "Hello"));
    invoker.run(std::make_unique<AppendCommand>(editor, ", World"));
    editor.print();

    invoker.undo();
    editor.print();

    invoker.undo();
    editor.print();
    return 0;
}
